packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,605 @@
+    using BytesUtil         for bytes;
+    using BlockReader       for bytes;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+
+    struct Context
+    {
+        bytes32 domainSeparator;
+        uint32  accountID;
+        uint    txIdx;
+    }
+
+    struct HashData
+    {
+        address                      connector;
+        bytes                        groupData;
+        TransferTransaction.Transfer transfer;
+        BridgeCall                   call;
+    }
+
+    bytes32 constant public BRIDGE_CALL_TYPEHASH = keccak256(
+        "BridgeCall(address from,address to,uint16 tokenID,uint96 amount,uint16 feeTokenID,uint96 maxFee,uint32 storageID,uint32 minGas,address connector,bytes groupData,bytes userData,uint256 validUntil)"
+    );
+
+    uint               public constant  MAX_NUM_TRANSACTIONS_IN_BLOCK = 386;
+    uint               public constant  MAX_AGE_PENDING_TRANSFERS     = 7 days;
+    uint               public constant  MAX_FEE_BIPS                  = 25;     // 0.25%
+    uint               public constant  GAS_LIMIT_CHECK_GAS_LIMIT     = 10000;
+
+    IExchangeV3        public immutable exchange;
+    uint32             public immutable accountID;
+    IDepositContract   public immutable depositContract;
+    bytes32            public immutable DOMAIN_SEPARATOR;
+
+
+    address                                    public exchangeOwner;
+
+    mapping (bytes32 => uint)                  public pendingTransfers;
+    mapping (bytes32 => mapping(uint => bool)) public withdrawn;
+
+    mapping (address => bool)                  public trustedConnectors;
+
+    uint                                       public batchIDGenerator;
+
+    // token -> tokenID
+    mapping (address => uint16)                public cachedTokenIDs;
+
+    event Transfers(uint batchID, InternalBridgeTransfer[] transfers);
+
+    event BridgeCallSuccess (bytes32 hash);
+    event BridgeCallFailed  (bytes32 hash, string reason);
+
+    event ConnectorTrusted  (address connector, bool trusted);
+
+    modifier onlyFromExchangeOwner()
+    {
+        require(msg.sender == exchangeOwner, "UNAUTHORIZED");
+        _;
+    }
+
+    constructor(
+        IExchangeV3      _exchange,
+        uint32           _accountID
+        )
+    {
+        exchange = _exchange;
+        accountID = _accountID;
+
+        depositContract = _exchange.getDepositContract();
+        exchangeOwner = _exchange.owner();
+
+        DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("Bridge", "1.0", address(this)));
+    }
+
+    function batchDeposit(
+        address                 from,
+        BridgeTransfer[] memory deposits
+        )
+        external
+        payable
+    {
+        require(from == msg.sender, "UNAUTHORIZED");
+        // Needs to be possible to do all transfers in a single block
+        require(deposits.length <= MAX_NUM_TRANSACTIONS_IN_BLOCK, "MAX_DEPOSITS_EXCEEDED");
```

Comment:
Currently the deposit itself does not need to be done through the bridge because it's just a normal deposit request.

There are points that could be made to let the deposit itself also to be done through the bridge, but that complicates things somewhat as well. The current method just puts the gas costs for all the expensive things to the caller of `batchDeposit`, the operator just needs to do the cheap L2 transfers.

Reply:
Do you think there is a problem if the deposit L2 tx is processed in a different block than all other transfers? The current implementation is just limited to do all L2 transfers, which is exactly `deposits.length` transfers, in the same block because the callback only works on a single block. But even that could be solved (but I don't think that will a problem in practice so not worth doing for now).